File: Sources/BidLabCore/Attribution.swift
import Foundation

/// Attribution credit allocation across an ordered set of touchpoints. Each
/// function returns a credit vector that sums to 1, so a single conversion is
/// distributed without inflating or losing credit.
public enum Attribution {
    /// All credit to the last touch.
    public static func lastTouch(_ n: Int) -> [Double] {
        guard n > 0 else { return [] }
        var v = Array(repeating: 0.0, count: n)
        v[n - 1] = 1
        return v
    }

    /// All credit to the first touch.
    public static func firstTouch(_ n: Int) -> [Double] {
        guard n > 0 else { return [] }
        var v = Array(repeating: 0.0, count: n)
        v[0] = 1
        return v
    }

    /// Equal credit across every touch.
    public static func linear(_ n: Int) -> [Double] {
        guard n > 0 else { return [] }
        return Array(repeating: 1.0 / Double(n), count: n)
    }

    /// U-shaped: the first and last touch each get `endsWeight`, the rest split
    /// the remainder evenly.
    public static func positionBased(_ n: Int, endsWeight: Double = 0.4) -> [Double] {
        guard n > 0 else { return [] }
        if n == 1 { return [1] }
        if n == 2 { return [0.5, 0.5] }
        var v = Array(repeating: 0.0, count: n)
        v[0] = endsWeight
        v[n - 1] = endsWeight
        let middle = (1 - 2 * endsWeight) / Double(n - 2)
        for i in 1..<(n - 1) { v[i] = middle }
        return v
    }
}

File: web/src/engine/attribution.ts
// Port of BidLabCore/Attribution.swift. Each function returns a credit vector
// over an ordered set of touchpoints that sums to 1 — one conversion distributed
// without inflating or losing credit.

/** All credit to the last touch. */
export function lastTouch(n: number): number[] {
  if (!(n > 0)) return []
  const v = new Array<number>(n).fill(0)
  v[n - 1] = 1
  return v
}

/** All credit to the first touch. */
export function firstTouch(n: number): number[] {
  if (!(n > 0)) return []
  const v = new Array<number>(n).fill(0)
  v[0] = 1
  return v
}

/** Equal credit across every touch. */
export function linear(n: number): number[] {
  if (!(n > 0)) return []
  return new Array<number>(n).fill(1 / n)
}

/** U-shaped: first and last touch each get `endsWeight`; the rest split the rest. */
export function positionBased(n: number, endsWeight = 0.4): number[] {
  if (!(n > 0)) return []
  if (n === 1) return [1]
  if (n === 2) return [0.5, 0.5]
  const v = new Array<number>(n).fill(0)
  v[0] = endsWeight
  v[n - 1] = endsWeight
  const middle = (1 - 2 * endsWeight) / (n - 2)
  for (let i = 1; i < n - 1; i++) v[i] = middle
  return v
}
